pytest 4.6.0 breaks matplotlib test gathering with `TypeError: cannot unpack non-iterable NoneType object`

This build: https://travis-ci.org/matplotlib/matplotlib/jobs/540086010 passed fine, with pytest 4.5.0. The next build from the same code, initiated via cron and using pytest 4.6.0: https://travis-ci.org/matplotlib/matplotlib/jobs/540212001 failed with

```
Calling pytest with the following arguments: -raR --maxfail=50 --timeout=300 --durations=25 --cov-report= --cov=lib -n 2 --log-level=DEBUG
============================= test session starts ==============================
platform linux -- Python 3.7.1, pytest-4.6.0, py-1.7.0, pluggy-0.12.0
rootdir: /home/travis/build/matplotlib/matplotlib, inifile: pytest.ini, testpaths: lib
plugins: forked-1.0.2, cov-2.7.1, rerunfailures-7.0, xdist-1.28.0, timeout-1.3.3, faulthandler-1.6.0
timeout: 300.0s
timeout method: signal
timeout func_only: False
gw0 I / gw1 I
gw0 [5784] / gw1 [5784]
...
_______ ERROR collecting lib/matplotlib/tests/test_constrainedlayout.py ________
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:963: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:902: in _find_spec
    ???
<frozen importlib._bootstrap>:876: in _find_spec_legacy
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/opt/python/3.7.1/lib/python3.7/ast.py:262: in visit
    return visitor(node)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
E   TypeError: cannot unpack non-iterable NoneType object
_______ ERROR collecting lib/matplotlib/tests/test_constrainedlayout.py ________
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
<frozen importlib._bootstrap>:983: in _find_and_load
    ???
<frozen importlib._bootstrap>:963: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:902: in _find_spec
    ???
<frozen importlib._bootstrap>:876: in _find_spec_legacy
    ???
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:169: in find_module
    source_stat, co = _rewrite_test(self.config, fn_pypath)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:404: in _rewrite_test
    rewrite_asserts(tree, fn, config)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:453: in rewrite_asserts
    AssertionRewriter(module_path, config).run(mod)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:701: in run
    new.extend(self.visit(child))
/opt/python/3.7.1/lib/python3.7/ast.py:262: in visit
    return visitor(node)
../../../virtualenv/python3.7.1/lib/python3.7/site-packages/_pytest/assertion/rewrite.py:828: in visit_Assert
    top_condition, explanation = self.visit(assert_.test)
E   TypeError: cannot unpack non-iterable NoneType object
____________ ERROR collecting lib/matplotlib/tests/test_mathtext.py ____________
In test_mathtext_rendering: 1 parameter sets specified, with different number of ids: 83
```

The operating system is Ubuntu 16.04.6 LTS (xenial).

I don't know how to unpack the line numbers for the breaking tests from the above, but [test_constrainedlayout.py](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/tests/test_constrainedlayout.py) has the following asserts:

```
    assert all(ax.get_position().extents == ax2.get_position().extents)
    assert all(ax.get_position().extents == ax2.get_position().extents)
    np.testing.assert_allclose(extents0, extents1)
    np.testing.assert_allclose(extents0, extents1)
```

In addition, there are multiple tests using matplotlib's `@image_comparison` decorator, but there are so many of those all over the codebase that I think then we'd see much more failures.

<!--
Thanks for submitting an issue!

Here's a quick checklist for what to provide:
-->

- [x] a detailed description of the bug or suggestion
- [ ] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [ ] minimal example if possible

--- a/changelog/5358.bugfix.rst
+++ b/changelog/5358.bugfix.rst
@@ -1,0 +1,1 @@
+Fix assertion rewriting of ``all()`` calls to deal with non-generators.

--- a/src/_pytest/assertion/rewrite.py
+++ b/src/_pytest/assertion/rewrite.py
@@ -949,11 +949,21 @@ warn_explicit(
         res = self.assign(ast.BinOp(left_expr, binop.op, right_expr))
         return res, explanation
 
+    @staticmethod
+    def _is_any_call_with_generator_or_list_comprehension(call):
+        """Return True if the Call node is an 'any' call with a generator or list comprehension"""
+        return (
+            isinstance(call.func, ast.Name)
+            and call.func.id == "all"
+            and len(call.args) == 1
+            and isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp))
+        )
+
     def visit_Call_35(self, call):
         """
         visit `ast.Call` nodes on Python3.5 and after
         """
-        if isinstance(call.func, ast.Name) and call.func.id == "all":
+        if self._is_any_call_with_generator_or_list_comprehension(call):
             return self._visit_all(call)
         new_func, func_expl = self.visit(call.func)
         arg_expls = []
@@ -980,8 +990,6 @@ warn_explicit(
 
     def _visit_all(self, call):
         """Special rewrite for the builtin all function, see #5062"""
-        if not isinstance(call.args[0], (ast.GeneratorExp, ast.ListComp)):
-            return
         gen_exp = call.args[0]
         assertion_module = ast.Module(
             body=[ast.Assert(test=gen_exp.elt, lineno=1, msg="", col_offset=1)]
@@ -1009,7 +1017,7 @@ warn_explicit(
         """
         visit `ast.Call nodes on 3.4 and below`
         """
-        if isinstance(call.func, ast.Name) and call.func.id == "all":
+        if self._is_any_call_with_generator_or_list_comprehension(call):
             return self._visit_all(call)
         new_func, func_expl = self.visit(call.func)
         arg_expls = []

--- a/testing/test_assertrewrite.py
+++ b/testing/test_assertrewrite.py
@@ -677,7 +677,7 @@ class TestAssertionRewrite(object):
         assert "UnicodeDecodeError" not in msg
         assert "UnicodeEncodeError" not in msg
 
-    def test_unroll_generator(self, testdir):
+    def test_unroll_all_generator(self, testdir):
         testdir.makepyfile(
             """
             def check_even(num):
@@ -692,7 +692,7 @@ class TestAssertionRewrite(object):
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
 
-    def test_unroll_list_comprehension(self, testdir):
+    def test_unroll_all_list_comprehension(self, testdir):
         testdir.makepyfile(
             """
             def check_even(num):
@@ -706,6 +706,31 @@ class TestAssertionRewrite(object):
         )
         result = testdir.runpytest()
         result.stdout.fnmatch_lines(["*assert False*", "*where False = check_even(1)*"])
+
+    def test_unroll_all_object(self, testdir):
+        """all() for non generators/non list-comprehensions (#5358)"""
+        testdir.makepyfile(
+            """
+            def test():
+                assert all((1, 0))
+            """
+        )
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines(["*assert False*", "*where False = all((1, 0))*"])
+
+    def test_unroll_all_starred(self, testdir):
+        """all() for non generators/non list-comprehensions (#5358)"""
+        testdir.makepyfile(
+            """
+            def test():
+                x = ((1, 0),)
+                assert all(*x)
+            """
+        )
+        result = testdir.runpytest()
+        result.stdout.fnmatch_lines(
+            ["*assert False*", "*where False = all(*((1, 0),))*"]
+        )
 
     def test_for_loop(self, testdir):
         testdir.makepyfile(

